File: frontend/src/api/reports.ts
/**
 * Reports and Exports API client functions.
 */

import { authFetch } from './auth'

const API_URL = import.meta.env.VITE_API_URL || 'http://localhost:8000'

// ============== Types ==============

export interface Report {
  id: string
  wallet_id: string | null
  wallet_name: string | null
  period_start: string
  period_end: string
  has_markdown: boolean
  has_pdf: boolean
  generated_by: 'manual' | 'scheduled'
  ai_model: string | null
  created_at: string
  updated_at: string
}

export interface ReportDetail extends Report {
  report_markdown: string | null
}

export interface ReportListResponse {
  reports: Report[]
  total: number
}

export interface GenerateReportRequest {
  wallet_id?: string
  period_start: string
  period_end: string
  include_ai_insights?: boolean
}

export interface ExportFilters {
  wallet_id?: string
  category_id?: string
  vendor_id?: string
  start_date?: string
  end_date?: string
  direction?: 'debit' | 'credit'
}

// ============== API Functions ==============

export async function fetchReports(params: {
  wallet_id?: string
  limit?: number
  offset?: number
} = {}): Promise<ReportListResponse> {
  const searchParams = new URLSearchParams()
  if (params.wallet_id) searchParams.set('wallet_id', params.wallet_id)
  if (params.limit) searchParams.set('limit', String(params.limit))
  if (params.offset) searchParams.set('offset', String(params.offset))

  const res = await authFetch(`${API_URL}/reports?${searchParams}`)
  if (!res.ok) throw new Error('Failed to fetch reports')
  return res.json()
}

export async function fetchReport(id: string): Promise<ReportDetail> {
  const res = await authFetch(`${API_URL}/reports/${id}`)
  if (!res.ok) throw new Error('Failed to fetch report')
  return res.json()
}

export async function generateReport(data: GenerateReportRequest): Promise<Report> {
  const res = await authFetch(`${API_URL}/reports`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(data),
  })
  if (!res.ok) throw new Error('Failed to generate report')
  return res.json()
}

export async function deleteReport(id: string): Promise<void> {
  const res = await authFetch(`${API_URL}/reports/${id}`, {
    method: 'DELETE',
  })
  if (!res.ok) throw new Error('Failed to delete report')
}

export function getReportPdfUrl(id: string): string {
  return `${API_URL}/reports/${id}/pdf`
}

export async function downloadReportPdf(id: string): Promise<Blob> {
  const res = await authFetch(`${API_URL}/reports/${id}/pdf`)
  if (!res.ok) throw new Error('Failed to download PDF')
  return res.blob()
}

// ============== Export Functions ==============

export async function exportTransactionsCsv(filters: ExportFilters = {}): Promise<Blob> {
  const searchParams = new URLSearchParams()
  if (filters.wallet_id) searchParams.set('wallet_id', filters.wallet_id)
  if (filters.category_id) searchParams.set('category_id', filters.category_id)
  if (filters.vendor_id) searchParams.set('vendor_id', filters.vendor_id)
  if (filters.start_date) searchParams.set('start_date', filters.start_date)
  if (filters.end_date) searchParams.set('end_date', filters.end_date)
  if (filters.direction) searchParams.set('direction', filters.direction)

  const res = await authFetch(`${API_URL}/reports/export/transactions.csv?${searchParams}`)
  if (!res.ok) throw new Error('Failed to export transactions')
  return res.blob()
}

export async function exportCategoriesCsv(params: {
  period_start: string
  period_end: string
  wallet_id?: string
}): Promise<Blob> {
  const searchParams = new URLSearchParams()
  searchParams.set('period_start', params.period_start)
  searchParams.set('period_end', params.period_end)
  if (params.wallet_id) searchParams.set('wallet_id', params.wallet_id)

  const res = await authFetch(`${API_URL}/reports/export/categories.csv?${searchParams}`)
  if (!res.ok) throw new Error('Failed to export categories')
  return res.blob()
}

export async function exportVendorsCsv(params: {
  period_start: string
  period_end: string
  wallet_id?: string
  limit?: number
}): Promise<Blob> {
  const searchParams = new URLSearchParams()
  searchParams.set('period_start', params.period_start)
  searchParams.set('period_end', params.period_end)
  if (params.wallet_id) searchParams.set('wallet_id', params.wallet_id)
  if (params.limit) searchParams.set('limit', String(params.limit))

  const res = await authFetch(`${API_URL}/reports/export/vendors.csv?${searchParams}`)
  if (!res.ok) throw new Error('Failed to export vendors')
  return res.blob()
}

// Helper to trigger download
export function downloadBlob(blob: Blob, filename: string): void {
  const url = URL.createObjectURL(blob)
  const a = document.createElement('a')
  a.href = url
  a.download = filename
  document.body.appendChild(a)
  a.click()
  document.body.removeChild(a)
  URL.revokeObjectURL(url)
}

File: backend/app/api/routes/reports.py
"""Report and Export API endpoints."""

from datetime import date
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query
from fastapi.responses import Response, StreamingResponse
from sqlalchemy.orm import Session

from app.api.deps import get_current_user
from app.db.models import User
from app.db.session import get_db
from app.schemas.analytics import ExportRequest
from app.schemas.report import (
    ReportDetailResponse,
    ReportGenerateRequest,
    ReportListResponse,
    ReportResponse,
)
from app.services.export import ExportService
from app.services.report import ReportService

router = APIRouter()


# ============== Reports ==============


@router.post("", response_model=ReportResponse, status_code=201)
async def generate_report(
    request: ReportGenerateRequest,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> ReportResponse:
    """
    Generate a new financial report for a period.

    Creates both markdown and PDF versions of the report.
    Optionally includes AI-generated insights if include_ai_insights is True.
    """
    service = ReportService(db)
    report = service.generate_report(
        request,
        include_ai_insights=request.include_ai_insights,
    )

    return ReportResponse(
        id=report.id,
        wallet_id=report.wallet_id,
        wallet_name=report.wallet.name if report.wallet else None,
        period_start=report.period_start,
        period_end=report.period_end,
        has_markdown=report.report_markdown is not None,
        has_pdf=report.report_pdf_blob is not None,
        generated_by=report.generated_by.value,
        ai_model=report.ai_model,
        created_at=report.created_at,
        updated_at=report.updated_at,
    )


@router.get("", response_model=ReportListResponse)
async def list_reports(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
    wallet_id: str | None = Query(None, description="Filter by wallet UUID"),
    limit: int = Query(20, ge=1, le=100),
    offset: int = Query(0, ge=0),
) -> ReportListResponse:
    """
    List generated reports.
    """
    service = ReportService(db)
    wallet_uuid = UUID(wallet_id) if wallet_id else None

    return service.list_reports(
        wallet_id=wallet_uuid,
        limit=limit,
        offset=offset,
    )


@router.get("/{report_id}", response_model=ReportDetailResponse)
async def get_report(
    report_id: UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> ReportDetailResponse:
    """
    Get a report by ID with markdown content.
    """
    service = ReportService(db)
    report = service.get_report(report_id)

    if not report:
        raise HTTPException(status_code=404, detail="Report not found")

    return ReportDetailResponse(
        id=report.id,
        wallet_id=report.wallet_id,
        wallet_name=report.wallet.name if report.wallet else None,
        period_start=report.period_start,
        period_end=report.period_end,
        report_markdown=report.report_markdown,
        has_pdf=report.report_pdf_blob is not None,
        generated_by=report.generated_by.value,
        ai_model=report.ai_model,
        created_at=report.created_at,
        updated_at=report.updated_at,
    )


@router.get("/{report_id}/pdf")
async def download_report_pdf(
    report_id: UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> Response:
    """
    Download a report as PDF.
    """
    service = ReportService(db)
    report = service.get_report(report_id)

    if not report:
        raise HTTPException(status_code=404, detail="Report not found")

    if not report.report_pdf_blob:
        raise HTTPException(status_code=404, detail="PDF not available for this report")

    # Format filename
    filename = f"report_{report.period_start.strftime('%Y%m')}.pdf"

    return Response(
        content=report.report_pdf_blob,
        media_type="application/pdf",
        headers={
            "Content-Disposition": f'attachment; filename="{filename}"',
        },
    )


@router.delete("/{report_id}", status_code=204)
async def delete_report(
    report_id: UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> None:
    """
    Delete a report.
    """
    service = ReportService(db)
    deleted = service.delete_report(report_id)

    if not deleted:
        raise HTTPException(status_code=404, detail="Report not found")


# ============== Exports ==============


@router.get("/export/transactions.csv")
async def export_transactions_csv(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
    wallet_id: str | None = Query(None),
    category_id: str | None = Query(None),
    vendor_id: str | None = Query(None),
    start_date: date | None = Query(None),
    end_date: date | None = Query(None),
    direction: str | None = Query(None, description="'debit' or 'credit'"),
) -> Response:
    """
    Export transactions to CSV.

    Filters are optional. Returns all transactions if no filters specified.
    """
    service = ExportService(db)

    request = ExportRequest(
        wallet_id=UUID(wallet_id) if wallet_id else None,
        category_id=UUID(category_id) if category_id else None,
        vendor_id=UUID(vendor_id) if vendor_id else None,
        start_date=start_date,
        end_date=end_date,
        direction=direction,
    )

    csv_content = service.export_transactions_csv(request)

    # Generate filename with date range if provided
    if start_date and end_date:
        filename = f"transactions_{start_date.strftime('%Y%m%d')}_{end_date.strftime('%Y%m%d')}.csv"
    else:
        filename = "transactions.csv"

    return Response(
        content=csv_content,
        media_type="text/csv",
        headers={
            "Content-Disposition": f'attachment; filename="{filename}"',
        },
    )


@router.get("/export/categories.csv")
async def export_categories_csv(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
    period_start: date = Query(..., description="Start date"),
    period_end: date = Query(..., description="End date"),
    wallet_id: str | None = Query(None),
) -> Response:
    """
    Export category spending summary to CSV.
    """
    service = ExportService(db)
    wallet_uuid = UUID(wallet_id) if wallet_id else None

    csv_content = service.export_category_summary_csv(
        period_start=period_start,
        period_end=period_end,
        wallet_id=wallet_uuid,
    )

    filename = f"category_summary_{period_start.strftime('%Y%m%d')}_{period_end.strftime('%Y%m%d')}.csv"

    return Response(
        content=csv_content,
        media_type="text/csv",
        headers={
            "Content-Disposition": f'attachment; filename="{filename}"',
        },
    )


@router.get("/export/vendors.csv")
async def export_vendors_csv(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
    period_start: date = Query(..., description="Start date"),
    period_end: date = Query(..., description="End date"),
    wallet_id: str | None = Query(None),
    limit: int = Query(50, ge=1, le=500),
) -> Response:
    """
    Export vendor spending summary to CSV.
    """
    service = ExportService(db)
    wallet_uuid = UUID(wallet_id) if wallet_id else None

    csv_content = service.export_vendor_summary_csv(
        period_start=period_start,
        period_end=period_end,
        wallet_id=wallet_uuid,
        limit=limit,
    )

    filename = f"vendor_summary_{period_start.strftime('%Y%m%d')}_{period_end.strftime('%Y%m%d')}.csv"

    return Response(
        content=csv_content,
        media_type="text/csv",
        headers={
            "Content-Disposition": f'attachment; filename="{filename}"',
        },
    )
